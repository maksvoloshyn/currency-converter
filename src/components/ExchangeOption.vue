<template>
    <div class="flex flex-col sm:flex-row w-full flex-auto">
        <currency-select
            class="sm:mr-2 mb-2 sm:mb-0"
            :code="code"
            :currencies="currencies"
            @change="updateCode"
        ></currency-select>
        <currency-input :disabled="!code" :amount="amount" @change="updateValue"></currency-input>
    </div>
</template>

<script>
    import CurrencyInput from '@/components/CurrencyInput';
    import CurrencySelect from '@/components/CurrencySelect';

    export default {
        name: 'CurrencyRow',

        components: {
            'currency-input': CurrencyInput,
            'currency-select': CurrencySelect,
        },

        props: {
            currencies: {
                type: Array,
                required: true,
            },

            amount: {
                type: Number,
                default: 0,
            },

            code: {
                type: String,
                default: null,
            },
        },

        methods: {
            updateValue(value) {
                this.$emit('update:amount', value);
            },

            updateCode(code) {
                this.$emit('update:code', code);
            },
        },
    };
</script>

<style scoped></style>
