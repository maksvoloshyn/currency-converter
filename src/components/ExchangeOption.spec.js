import {shallowMount} from '@vue/test-utils';
import ExchangeOption from '@/components/ExchangeOption';

const CurrencySelectStub = {
    template: '<span>{{code}}</span>',
    props: ['currencies', 'code'],
};

const CurrencyInputStub = {
    template: '<span>{{amount}}</span>',
    props: ['currencies', 'amount'],
};

function factory() {
    return shallowMount(ExchangeOption, {
        propsData: {
            currencies: currenciesMock,
            amount: 12,
            code: 'EUR',
        },
        stubs: {
            'currency-input': CurrencyInputStub,
            'currency-select': CurrencySelectStub,
        },
    });
}

const currenciesMock = [{UAH: 'Ukrainian hryvna', EUR: 'Euro', USD: 'United States Dollar'}];

describe('ExchangeOption', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = factory();
    });

    it('renders correct template', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    describe('when the code is null', () => {
        it('sets disabled=true on CurrencyInput', async () => {
            await wrapper.setProps({code: null});

            expect(wrapper.getComponent(CurrencyInputStub).attributes('disabled')).toBe('disabled');
        });
    });

    describe('when CurrencyInput emits `change` event', () => {
        it('emits `update:amount` with a given value', async () => {
            await wrapper.getComponent(CurrencyInputStub).vm.$emit('change', 1033);

            expect(wrapper.emitted('update:amount')[0][0]).toBe(1033);
        });
    });

    describe('when updateCode emits `change` event', () => {
        it('emits `update:code` with a given value', async () => {
            await wrapper.getComponent(CurrencySelectStub).vm.$emit('change', 'USD');

            expect(wrapper.emitted('update:code')[0][0]).toBe('USD');
        });
    });
});
