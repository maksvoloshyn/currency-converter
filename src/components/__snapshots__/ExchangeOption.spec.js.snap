// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeOption renders correct template 1`] = `
<div
  class="flex flex-col sm:flex-row w-full flex-auto"
>
  <span
    class="sm:mr-2 mb-2 sm:mb-0"
  >
    EUR
  </span>
   
  <span
    aria-label="Amount of EUR"
  >
    12
  </span>
</div>
`;
